packages-next/fields-document/ARCHITECTURE.md

```diff
@@ -0,0 +1,107 @@
+# Document editor architecture
+
+The document editor uses [Slate](https://github.com/ianstormtaylor/slate). To learn about Slate generally, go through https://docs.slatejs.org/walkthroughs and https://docs.slatejs.org/concepts. The sections on [Nodes](https://docs.slatejs.org/concepts/02-nodes) and [Locations](https://docs.slatejs.org/concepts/03-locations) contain especially important terminology.
+
+## Data storage
+
+The document field type stores its data as a JSON blob, representing the Slate data.
+A trivial document with a single heading and paragraph would look like:
+
+```json
+[
+  {
+    "type": "heading",
+    "level": 1,
+    "children": [
+      {
+        "text": "content"
+      }
+    ]
+  }
+  {
+    "type": "paragraph",
+    "children": [
+      {
+        "text": "some text"
+      }
+    ]
+  }
+]
+```
+
+## Plugins
+
+For normalization and handling user input and some other things(you can see what they are by looking at the type definition for the `Editor` in Slate), you write [Slate "plugins"](https://docs.slatejs.org/concepts/07-plugins).
+Plugins are not some special API, it means accepting an `Editor` object and returning it but modifying some of the properties.
+
+## Rendering
+
+The rendering of the editor is like any other bit of React code.
+See https://docs.slatejs.org/concepts/08-rendering for how rendering the document works in Slate.
+
+The editor is very performance-sensitive because it will re-render on every keystroke so keep these things in mind while writing code for rendering the editor:
+
+- Renderers for elements/leaves should never use `useSlate` or `useToolbarState` because they will change on every editor change. If they need to get the editor to do something in response to a user action, they should use `useStaticEditor` which will not cause re-renders.
```

Comment:
```suggestion
- Renderers for elements/leaves should never use `useSlate` or `useToolbarState` because those hooks will cause a re-render on every editor change. If they need to get the editor to do something in response to a user action, they should use `useStaticEditor` which will not cause re-renders.
```